benchmark/QR_accuracy_stability_analysis.cc

```diff
@@ -15,148 +15,317 @@
 
 template <typename T>
 static void
-test_cond_helper_0(int64_t m, 
+error_check(int64_t m,
+            int64_t n,
+            int64_t k,
+            T norm_A,
+            T* A_dat,
+            T* A_1_dat,
+            T* Q_dat,
+            T* R_dat,
+            T* I_ref_dat
+            ) {
+
+    // Check orthogonality of Q
+    // Q' * Q  - I = 0
+    blas::gemm(Layout::ColMajor, Op::Trans, Op::NoTrans, k, k, m, 1.0, Q_dat, m, Q_dat, m, -1.0, I_ref_dat, k);
+    T norm_0 = lapack::lange(lapack::Norm::Fro, k, k, I_ref_dat, k);
+
+    // AP - QR
+    blas::gemm(Layout::ColMajor, Op::NoTrans, Op::NoTrans, m, n, k, 1.0, Q_dat, m, R_dat, k, -1.0, A_dat, m);
+    
+    // Implementing max col norm metric
+    T max_col_norm = 0.0;
+    T col_norm = 0.0;
+    int max_idx = 0;
+    for(int i = 0; i < n; ++i) {
+        col_norm = blas::nrm2(m, A_dat + (m * i), 1);
+        if(max_col_norm < col_norm) {
+            max_col_norm = col_norm;
+            max_idx = i;
+        }
+    }
+    T col_norm_A = blas::nrm2(n, A_1_dat + (m * max_idx), 1);
+    T norm_AQR = lapack::lange(Norm::Fro, m, n, A_dat, m);
+    
+    printf("REL NORM OF AP - QR: %15e\n", norm_AQR / norm_A);
+    printf("MAX COL NORM METRIC: %15e\n", max_col_norm / col_norm_A);
+    printf("FRO NORM OF Q' * Q - I: %2e\n\n", norm_0);
+}
+
+
+template <typename T>
+static void
+cholqr_helper(int64_t m, 
                   int64_t n, 
                   const std::tuple<int, T, bool>& mat_type, 
                   RandBLAS::base::RNGState<r123::Philox4x32> state) {
 
     std::vector<T> A(m * n, 0.0);
+    std::vector<T> A_hat(m * n, 0.0);
     std::vector<T> I_ref(n * n, 0.0);
+    std::vector<T> R_sp(n * n, 0.0);
 
     // Generate random matrix
     RandLAPACK::util::gen_mat_type(m, n, A, n, state, mat_type);
     RandLAPACK::util::eye(n, n, I_ref);
 
-    // CHOL QR
-    RandLAPACK::CholQRQ<T> Orth_CholQR(false, false);
-    // Orthonormalize A
-    Orth_CholQR.call(m, n, A);
+    std::copy(A.data(), A.data() + (m * n), A_hat.data());
 
     T* A_dat = A.data();
+    T* A_hat_dat = A_hat.data();
     T* I_ref_dat = I_ref.data();
+    T* R_sp_dat  = R_sp.data();
+
+    blas::syrk(Layout::ColMajor, Uplo::Upper, Op::Trans, n, m, 1.0, A_dat, m, 0.0, R_sp_dat, n);
+    lapack::potrf(Uplo::Upper, n, R_sp_dat, n);
+    blas::trsm(Layout::ColMajor, Side::Right, Uplo::Upper, Op::NoTrans, Diag::NonUnit, m, n, 1.0, R_sp_dat, n, A_dat, m);
 
     // Check orthogonality of Q
     // Q' * Q  - I = 0
     blas::gemm(Layout::ColMajor, Op::Trans, Op::NoTrans, n, n, m, 1.0, A_dat, m, A_dat, m, -1.0, I_ref_dat, n);
     T norm_Q = lapack::lange(lapack::Norm::Fro, n, n, I_ref_dat, n);
+    blas::gemm(Layout::ColMajor, Op::NoTrans, Op::NoTrans, m, n, n, 1.0, A_dat, m, R_sp_dat, n, -1.0, A_hat_dat, m);
+    T norm_A = lapack::lange(Norm::Fro, m, n, A_hat_dat, m);
```

Comment:
More examples where you should use ``syrk`` instead of ``gemm``.

Reply:
Fixed.